lib/puppet/provider/elasticsearch_plugin/plugin.rb

```diff
@@ -0,0 +1,157 @@
+$LOAD_PATH.unshift(File.join(File.dirname(__FILE__),"..","..",".."))
+
+Puppet::Type.type(:elasticsearch_plugin).provide(:plugin) do
+  desc "A provider for the resource type `elasticsearch_plugin`,
+        which handles plugin installation"
+
+  commands :plugin => '/usr/share/elasticsearch/bin/plugin'
+  commands :es => '/usr/share/elasticsearch/bin/elasticsearch'
+
+  def exists?
+    es_version
+    if !File.exists?(pluginfile)
+      debug "Plugin file #{pluginfile} does not exist"
+      return false
+    elsif File.exists?(pluginfile) && readpluginfile != pluginfile_content
+      debug "Got #{readpluginfile} Expected #{pluginfile_content}"
+      self.destroy
+      false
+    else
+      debug "Plugin exists"
+      return true
+    end
+  end
+
+  def pluginfile_content
+    if is2x?
+      items = @resource[:name].split("/")
+      if items.count == 1 # Official plugin
+        version = plugin_version(@resource[:name])
+        return "#{@resource[:name]}/#{version}"
+      else
+        return @resource[:name]
+      end
+    else
+      return @resource[:name]
+    end
+  end
+
+  def pluginfile
+    File.join(@resource[:plugin_dir], plugin_name(@resource[:name]), '.name')
+  end
+
+  def writepluginfile
+    File.open(pluginfile, 'w') do |file|
+      file.write pluginfile_content
+    end
+  end
+
+  def readpluginfile
+    f = File.open(pluginfile)
+    f.readline
+  end
+
+  def install1x
+    if !@resource[:url].nil?
+      commands = [ plugin_name(@resource[:name]), '--url', @resource[:url] ]
+    elsif !@resource[:source].nil?
+      commands = [ plugin_name(@resource[:name]), '--url', "file://#{@resource[:source]}" ]
+    else
+      commands = [ @resource[:name] ]
+    end
+    commands
+  end
+
+  def install2x
+    if !@resource[:url].nil?
+      commands = [ @resource[:url] ]
+    elsif !@resource[:source].nil?
+      commands = [ "file://#{@resource[:source]}" ]
+    else
+      commands = [ @resource[:name] ]
+    end
+    commands
```

Comment:
imo: there should be only one source, and that source can alternatively be a URL, or a filesystem path.
if it is the latter, but doesn't start with `file://`, we make it start with `file://`

that way, these two functions are simplified reduced to only two cases each.